service-wcs/src/main/java/org/oskari/wcs/geotiff/IFD.java

```diff
@@ -0,0 +1,383 @@
+    private int sMinSampleValue;
+    private int sMaxSampleValue;
+
+    private int[] geoKeyDirectory;
+    private double[] modelTransformation;
+
+
+    public IFD(ByteBuffer bb, int pos) {
+        // Set the initial position
+        bb.position(pos);
+        final int tagCount = bb.getShort() & 0xffff;
+        for (int i = 0; i < tagCount; i++) {
+            // Mark current position
+            pos = bb.position();
+            // id and type are uint16
+            int id = bb.getShort() & 0xFFFF;
+            int type = bb.getShort() & 0xFFFF;
+            int count = bb.getInt();
+            int offset = bb.getInt();
+            parseTag(bb, id, type, count, offset);
+            // Move to next position
+            bb.position(pos + 12);
+        }
+    }
+
+    private void parseTag(ByteBuffer bb, int id, int type, int count, int offset) {
+        int typeLen = getFieldTypeLength(type);
+        int valueSize = typeLen * count;
+        if (valueSize <= 4) {
+            int value;
+            if (bb.order() == ByteOrder.BIG_ENDIAN) {
+                int shift = (4 - typeLen) * 8;
+                value = offset >>> shift;
+            } else {
+                value = offset;
+            }
+            setValue(id, value);
+        } else {
+            bb.position(offset);
+            parseValue(bb, id, type, count);
+        }
+    }
+
+    private void parseValue(ByteBuffer bb, int id, int type, int count) {
+        if (count == 1) {
+            parseSingleValue(bb, id, type);
+        } else {
+            parseMultipleValues(bb, id, type, count);
+        }
+    }
+
+    private void parseSingleValue(ByteBuffer bb, int id, int type) {
+        // Ignore types of <= 4 bytes
+        switch (type) {
+        case 5:
+        case 10:
+            setValue(id, bb.getLong());
+            break;
+        }
+    }
+
+    private void parseMultipleValues(ByteBuffer bb, int id, int type, int count) {
+        switch (type) {
+        case 3:
+            setValue(id, getUShortArray(bb, count));
+            break;
+        case 4:
+        case 9:
+            setValue(id, getIntArray(bb, count));
+            break;
+        case 12:
+            setValue(id, getDoubleArray(bb, count));
+            break;
+        }
+    }
+
+    private static int[] getUShortArray(ByteBuffer bb, int count) {
+        int[] arr = new int[count];
+        for (int i = 0; i < count; i++) {
+            arr[i] = bb.getShort() & 0xFFFF;
+        }
+        return arr;
+    }
+
+    private static int[] getIntArray(ByteBuffer bb, int count) {
+        int[] arr = new int[count];
+        bb.asIntBuffer().get(arr);
+        return arr;
+    }
+
+    private static double[] getDoubleArray(ByteBuffer bb, int count) {
+        double[] arr = new double[count];
+        bb.asDoubleBuffer().get(arr);
+        return arr;
+    }
+
+    private static final int getFieldTypeLength(int type) throws IllegalArgumentException {
```

Comment:
Some method comment on this would help figure out what's going on.

Reply:
Added a comment for each of the types